internal/testutils/testserver.go

```diff
@@ -56,9 +62,31 @@ type ServerResponses map[HTTPMethod]struct {
 }
 
 // NewHTTPTestServer creates a new HTTP test server with the specified responses for each HTTP method.
-func NewHTTPTestServer(t *testing.T, arg ServerResponses) *TestServer {
-	return &TestServer{Server: httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
-		if res, found := arg[req.Method]; found {
+func NewHTTPTestServer(t *testing.T, responses []ServerResponses) *TestServer {
+	testServer := &TestServer{}
+	handler := func(rw http.ResponseWriter, req *http.Request) {
+		testServer.calls++
+		// Special case: if only one response is specified, this will be reused
+		// for each and every call
+		if len(responses) == 1 {
+			if res, found := responses[0][req.Method]; found {
+				if res.ValidateRequestFunc != nil {
+					res.ValidateRequestFunc(req)
+				}
+				rw.WriteHeader(res.ResponseCode)
+				_, _ = rw.Write([]byte(res.ResponseBody)) // nosemgrep: go.lang.security.audit.xss.no-direct-write-to-responsewriter.no-direct-write-to-responsewriter
+			} else {
+				t.Errorf("unexpected HTTP method call: %s", req.Method)
+			}
+			return
+		}
+
+		if len(responses) <= testServer.calls-1 {
+			t.Fail()
+			t.Fatalf("Exceeded number of calls to test server (expected: %d)", len(responses))
```

Comment:
I think `Fatal/Fatalf` already call a `FailNow`(no guarantee), so the t.Fail() on line 85 isn't needed

Reply:
Also general note: I think this server should be consistent on when it stops a test. 
As is an unexpected HTTP method will fail the test, but let it continue (Errorf), but a totally unexpected call will fail the test immediately (Fatalf). 

I don't have a strong preference (generally I prefer stopping, but failing and continuing seems to be the more usual default), but I think this should use Errorf or Fatalf exclusively.